contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -103,20 +128,126 @@ fn ensure_its_source_address(
     Ok(())
 }
 
+fn verify_coin(
+    deps: &DepsMut,
+    coin: &Option<Coin>,
+    message: &Message,
+    source_chain: &ChainNameRaw,
+) -> Result<(), Error> {
+    match coin {
+        None => verify_coin_should_be_none(deps, message, source_chain),
```

Comment:
I recommend inverting the checks.
1. Check if source chain is a core chain. If yes,
  a. Check if token is a gateway token
  b. Otherwise, coin should be empty
2. Otherwise, coin should be empty
1b. and 2. could be combined as the else case potentially. Maybe something like

```
verify_coin(..) {
  let transfer_coin: Option<Coin> = gateway_token_transfer(source_chain, message); 
  ensure!(transfer_coin == coin, ..);
}
```
`gateway_token_transfer` checks if source chain is core, and if token is gateway denom, and returns a coin representing the expected transfer_token/transfer_coin